File: backend/src/modules/backup/backup.contract.ts
/**
 * O contrato de backup entre o Spark Android e o Spark Backend (T16.4).
 *
 * A definição legível — envelope, identidades, hash, idempotência, erros — vive em
 * `contracts/backup/v1/README.md`, junto das fixtures que os testes dos dois lados consomem.
 * Este arquivo é a metade TypeScript dela: os nomes que não podem divergir em silêncio.
 *
 * O espelho Kotlin é `com.example.data.backup.BackupContract`.
 */

/**
 * Versão do **formato de backup**.
 *
 * Não se confunde com nenhuma das outras versões do projeto: não é a versão do banco Room, não é
 * o `/v1` da API, não é `entitySchemaVersion` (que é por agregado) e não é a versão do app.
 */
export const BACKUP_SCHEMA_VERSION = 1;

export const SUPPORTED_BACKUP_SCHEMA_VERSIONS: readonly number[] = [1];

export function isSupportedBackupSchemaVersion(version: number): boolean {
  return SUPPORTED_BACKUP_SCHEMA_VERSIONS.includes(version);
}

/**
 * Os agregados que um backup pode conter. Registry **fechado**.
 *
 * Os seis primeiros são os `SyncEntityType` da T16.3. Os três últimos existem só no backup: eles
 * são dado pessoal do Grupo A da matriz, mas ainda não produzem mutação incremental — o snapshot
 * completo os cobre, e a T16.6 precisará lhes dar mutação própria.
 */
export const BACKUP_ENTITY_TYPES = [
  'WORKOUT_PROGRAM',
  'WORKOUT_TEMPLATE',
  'WORKOUT_SESSION',
  'CUSTOM_EXERCISE',
  'BODY_MEASUREMENT',
  'CHECK_IN',
  'EXERCISE_OVERRIDE',
  'WEEKLY_GOAL',
  'USER_PREFERENCES',
] as const;

export type BackupEntityType = (typeof BACKUP_ENTITY_TYPES)[number];

/**
 * Códigos de erro do backup, no envelope da T16.0 (`{ error: { code, message, requestId } }`).
 *
 * Nenhuma mensagem associada a estes códigos repete conteúdo do snapshot — nem nome de treino,
 * nem nota, nem medida (§50 da T16.2, §77 da T16.4).
 */
export const BACKUP_ERROR_CODES = {
  /** Envelope, item, identidade ou relação fora do contrato. */
  INVALID_BACKUP: 'INVALID_BACKUP',
  /** `backupSchemaVersion` que este servidor não sabe interpretar. */
  UNSUPPORTED_BACKUP_SCHEMA_VERSION: 'UNSUPPORTED_BACKUP_SCHEMA_VERSION',
  /** `entitySchemaVersion` desconhecida para aquele `entityType`. */
  UNSUPPORTED_ENTITY_SCHEMA_VERSION: 'UNSUPPORTED_ENTITY_SCHEMA_VERSION',
  /** Mesmo `clientBackupId` da mesma conta, com conteúdo diferente. */
  BACKUP_IDEMPOTENCY_CONFLICT: 'BACKUP_IDEMPOTENCY_CONFLICT',
  /** Corpo, número de itens ou item acima do teto. */
  BACKUP_TOO_LARGE: 'BACKUP_TOO_LARGE',
  /** Nenhum backup desta conta — ou nenhum com aquele `backupId`, que é a mesma resposta. */
  BACKUP_NOT_FOUND: 'BACKUP_NOT_FOUND',
  /**
   * O snapshot existe, mas o servidor não tem o documento original dele (T16.5, T18.1).
   *
   * Acontece com backups criados antes de `0004_backup_payload.sql`, e — desde a T18.1 — quando
   * o objeto do Object Storage não existe ou não passa na verificação de integridade (tamanho e
   * SHA-256 contra a metadata). Eles continuam válidos como metadata e não podem ser restaurados:
   * devolver uma reconstrução, ou um documento cujo hash não fecha, seria pior do que dizer que
   * não dá.
   */
  BACKUP_CONTENT_UNAVAILABLE: 'BACKUP_CONTENT_UNAVAILABLE',
  /**
   * O Object Storage não respondeu (T18.1 §40): timeout, permissão, quota, rede.
   *
   * `503`, e não `410`: a diferença entre "este backup não pode ser restaurado" e "o servidor não
   * consegue alcançar o armazenamento agora" é a diferença entre desistir e tentar de novo. O
   * Android já trata `5xx` como indisponibilidade recuperável — no upload, a tentativa continua
   * pendente e o reenvio é idempotente; no download, nada local é alterado.
   */
  BACKUP_STORAGE_UNAVAILABLE: 'BACKUP_STORAGE_UNAVAILABLE',
  /**
   * Requisições de backup demais para esta conta em uma janela curta (T16.8).
   *
   * Não é perda: a tentativa de backup do aparelho é durável e imutável, e reenviá-la mais tarde
   * com o mesmo `clientBackupId` é idempotente por contrato. O que a recusa impede é um cliente em
   * laço transformar um app pessoal em carga contínua sobre a VPS.
   */
  BACKUP_RATE_LIMITED: 'BACKUP_RATE_LIMITED',
} as const;

export type BackupErrorCode = (typeof BACKUP_ERROR_CODES)[keyof typeof BACKUP_ERROR_CODES];

/**
 * O que o Android recebe de volta: **metadata**, nunca o snapshot.
 *
 * Devolver o conteúdo depois do upload não serviria a nada nesta fase e já seria meio caminho do
 * download que a T16.5 vai desenhar com validação e preview.
 */
export interface BackupMetadataResponse {
  readonly backupId: string;
  readonly clientBackupId: string;
  readonly backupSchemaVersion: number;
  /** Relógio do **servidor**. É esta a autoridade de "quando" e de "qual é o mais recente". */
  readonly createdAt: number;
  readonly itemCount: number;
  readonly sizeBytes: number;
  readonly payloadHash: string;
}

/**
 * A lista de backups retidos da conta autenticada (T16.5).
 *
 * Só metadata, e a ordem é do **servidor**: do mais recente para o mais antigo, por sequência
 * interna. O cliente não reordena por `createdAt` nem por relógio de aparelho.
 */
export interface BackupListResponse {
  readonly items: readonly BackupMetadataResponse[];
}

File: app/src/main/java/com/example/data/backup/BackupContract.kt
package com.example.data.backup

import com.example.data.sync.SyncEntityType

/**
 * O contrato de backup entre o Spark Android e o Spark Backend (T16.4).
 *
 * A definição legível — envelope, identidades, hash, idempotência, erros e tetos — vive em
 * `contracts/backup/v1/README.md`, junto das fixtures que os testes dos **dois** lados consomem.
 * Este arquivo é a metade Kotlin dela; a metade TypeScript é
 * `backend/src/modules/backup/backup.contract.ts`.
 *
 * ## Backup não é sincronização
 *
 * ```text
 * T16.4   Android ──snapshot completo──▶ Spark Backend      (existe)
 *         Android ◀───────────────────── Spark Backend      (não existe)
 * ```
 *
 * Um backup é uma cópia completa e **autocontida** do estado pessoal atual. Ele não depende de
 * backup anterior, de delta, da Outbox nem de servidor antigo. Restore é T16.5; push/pull
 * incremental é T16.6.
 */
object BackupContract {

    /**
     * Versão do **formato de backup**.
     *
     * Não se confunde com nenhuma das outras versões do projeto: não é `AppDatabase.version`, não
     * é o `/v1` da API, não é [BackupEntityType.schemaVersion] (que é por agregado) e não é a
     * versão do app.
     */
    const val SCHEMA_VERSION: Int = 1

    /** O caminho do endpoint de criação. Relativo à base do `SparkBackendClient`. */
    const val BACKUPS_PATH: String = "v1/backups"

    /** O caminho da metadata do backup mais recente. */
    const val LATEST_BACKUP_PATH: String = "v1/backups/latest"

    /** Identidade singleton das preferências sincronizáveis. */
    const val PREFERENCES_IDENTITY: String = "preferences"

    /** Prefixo da identidade natural de uma meta semanal. */
    const val WEEK_IDENTITY_PREFIX: String = "week:"

    /** Prefixo da identidade derivada de uma customização sobre exercício de catálogo. */
    const val CANONICAL_IDENTITY_PREFIX: String = "canonical:"

    /** Prefixo da identidade derivada de uma customização sobre exercício criado pelo usuário. */
    const val CUSTOM_IDENTITY_PREFIX: String = "custom:"
}

/**
 * Os agregados que um backup pode conter. Registry **fechado** dos dois lados.
 *
 * Os seis primeiros são exatamente os [SyncEntityType] da T16.3, e são montados pelo **mesmo**
 * `SyncAggregateSnapshotBuilder` — não existe um segundo serializador de treino no Spark.
 *
 * Os três últimos existem só no backup. Eles são dado pessoal do Grupo A da matriz, mas ainda não
 * produzem mutação incremental: a matriz da T16.3 já os marcava como "T16.4", e o snapshot
 * completo os cobre. Quando a T16.6 trouxer push incremental, eles precisarão de mutação própria —
 * está registrado como pendência em `ARCHITECTURE.md`.
 */
enum class BackupEntityType(
    /** A versão do payload **daquele agregado**, independente do envelope. */
    val schemaVersion: Int,
    /** O agregado equivalente da Outbox, quando existe. */
    val syncEntityType: SyncEntityType?
) {
    WORKOUT_PROGRAM(1, SyncEntityType.WORKOUT_PROGRAM),
    WORKOUT_TEMPLATE(1, SyncEntityType.WORKOUT_TEMPLATE),
    WORKOUT_SESSION(1, SyncEntityType.WORKOUT_SESSION),
    CUSTOM_EXERCISE(1, SyncEntityType.CUSTOM_EXERCISE),
    BODY_MEASUREMENT(1, SyncEntityType.BODY_MEASUREMENT),
    CHECK_IN(1, SyncEntityType.CHECK_IN),

    /** Identidade derivada do exercício alvo — ver `identity-contract.md`. */
    EXERCISE_OVERRIDE(1, null),

    /** Identidade natural: a semana de vigência. */
    WEEKLY_GOAL(1, null),

    /** Singleton por conta. Vem do DataStore, com DTO próprio. */
    USER_PREFERENCES(1, null);

    companion object {

        /** Os agregados que também têm identidade de Outbox, na ordem do registry. */
        val syncAggregates: List<BackupEntityType> = entries.filter { it.syncEntityType != null }
    }
}
